packages/@uppy/utils/src/fileFilters.js

```diff
@@ -0,0 +1,10 @@
+export function filterNonFailedFiles (files) {
+  const hasError = (file) => 'error' in file && file.error
+
+  return files.filter((file) => !hasError(file))
+}
+
+// Don't double-emit upload-started for Golden Retriever-restored files that were already started
+export function filterFilesToEmitUploadStarted (files) {
+  return files.filter((file) => !file.progress.uploadStarted || !file.isRestored)
+}
```

Comment:
> the reason why I wanted to make the util is not for performance

I assumed you don't want to use `getFilesObjectFromState` because it does more than you need and therefor does unnecessary computations. If that's not the case, I would bring it into that function so we have one place to reason about, which will be handy when we want to do a bigger state refactor as has been brought up sometimes.

Reply:
by `getFilesObjectFromState`, do you mean `checkAndCreateFileStateObject`? (I renamed that function to `checkAndUpdateFileState` btw.)
I didn't put the `filterFilesToEmitUploadStarted` logic inside core.#checkAndUpdateFileState because the logic was never in core in the first place.

core.#checkAndUpdateFileState is called when `addFile` is called, but 
```
    const filesToEmit = filterFilesToEmitUploadStarted(filesFiltered)
    this.uppy.emit('upload-start', filesToEmit)
```
...is called when the upload starts. Adding files and uploading are different things, or I probably misunderstood what you meant.

Reply:
This is what I mean

https://github.com/transloadit/uppy/blob/f84ba514c1e69795150aef385ea1319ad53caccb/packages/%40uppy/core/src/Uppy.js#L326-L361

Reply:
gotcha, now i understand.  we were talking about different methods! Though I don't think it makes sense to put this filter inside of that method. then when i need that one filter, that would be a case of:
> I wanted a banana but what I got was a gorilla holding the banana and the entire jungle

...that was a bit exaggerated, but yea I think it's not optimal performance-wise to loop through `files` 9 times just to get 1 of the filtered variables, so we should probably split up `getObjectOfFilesPerState` into multiple getters or functions